logstash-core/lib/logstash/logging/slow_logger.rb

```diff
@@ -0,0 +1,68 @@
+# encoding: utf-8
+require "logstash/namespace"
+require "logstash/logging/slow_logger/freq_items"
+
+# This module contains necessary classes to handle slow log
+# operations.
+module LogStash; module Logging
+
+  # Slow log null logger, this logger does
+  # nothing when used.
+  class NullLogger
+    def log(event, threshold, time, data={})
+    end
+    alias_method :warn, :log
+  end
+
+  # Generic slow logger, this class is responsible of
+  # reporting to the specific logger and holding an
+  # aggregated view of it's reported events.
+  class SlowLogger
+
+    attr_reader :logger, :freq_items, :settings
+
+    def initialize(name="loggers.slow", settings=LogStash::SETTINGS, params={})
+      @logger     = LogStash::Logging::Logger.new(name)
+      @freq_items = LogStash::Logging::Util::FreqItems.new
+      @settings   = settings
+    end
+
+    def log(event, threshold, took_in_seconds, data={})
```

Comment:
it would be nice to have another method `log(event, threshold, start_millis, end_millis, data={})` that plugins can use. This way, plugins don't have to do the `elapsed = end - start` calculation and we can keep it DRY.